### ISSUE TITLE
[Website Bug]: Content Correction for https://unicorn-utterances.com/posts/ffg-fundamentals-accessing-children

### ISSUE THREAD

**Issue by joezimjs** [unicorn-utterances/unicorn-utterances#1073]
### Description

I'm a senior level Vue developer and I was just skimming through your book and found that Vue was stated as being unable to do a few things with their slots on https://unicorn-utterances.com/posts/ffg-fundamentals-accessing-children. I believe I can correct these examples:

## Counting a Component's Children

```vue
<!-- ParentList.vue -->
<template>
  <p>
    There are {{ $slots.default?.()?.length ?? 0 }} number of items in this
    array
  </p>
  <ul>
    <slot></slot>
  </ul>
</template>
```

And then you can use App.vue just like you did with React instead of need to pass a prop.

## Rendering Children in a Loop

```vue
<!-- ParentList.vue -->
<script setup lang="ts">
import { useSlots, computed } from 'vue';
const slots = useSlots();

// We only want to call the slot function once per render, so put it in a computed and use that value everywhere else
const defaultSlot = computed(() => slots.default?.());
</script>

<template>
  <p>There are {{ defaultSlot?.length ?? 0 }} items passed</p>
  <ul>
    <li v-for="child in defaultSlot"><component :is="child" /></li>
  </ul>
</template>
```

```vue
<!-- App.vue -->
<script setup lang="ts">
import ParentList from './ParentList.vue';
</script>

<template>
  <ParentList>
    <span style="color: red">Item 1</span>
    <span style="color: green">Item 2</span>
    <span style="color: blue">Item 3</span>
  </ParentList>
</template>
```

## Adding Children Dynamically

If you take the previous example and use a v-for on a single span, the ParentList will only see a single "wrapper" node in that slot that contains all the children. You CAN throw in some logic to detect this and work around it, but it requires you to mess with undocumented internals, so we'll avoid it. You can, however, go the route you've been going with some of your examples with passing the list in as a prop, which will make it nice and simple:

```vue
<!-- ParentList.vue -->
<script setup lang="ts">
import { useSlots, computed } from 'vue';

const props = defineProps({ list: Array });
</script>

<template>
  <p>There are {{ list?.length ?? 0 }} items passed</p>
  <ul>
    <li v-for="(item, i) in list"><slot v-bind="{ item, i }" /></li>
  </ul>
</template>
```

```vue
<!-- App.vue -->
<script setup lang="ts">
import { ref } from 'vue';
import ParentList from './ParentList.vue';

const list = ref([1, 42, 13]);

const addOne = () => {
  // `Math` is built into the browser
  const randomNum = Math.floor(Math.random() * 100);
  list.value.push(randomNum);
};
</script>

<template>
  <ParentList :list="list">
    <template #default="{ i, item }">
      <span>{{ i }} {{ item }}</span>
    </template>
  </ParentList>

  <button @click="addOne">Add</button>
</template>
```

The odd thing is that this is almost exactly the same as your example for the next section except we're not doing anything with background color and I put the &lt;li> in ParentList instead of in App, so I'm a bit surprised you didn't think this was possible.

**Comment by crutchcorn** [comment]
I'm currently investigating if `slots.default?.()?.length` has the same problems with children's `watch` and `watchEffect` that I have in my notes.

> so I'm a bit surprised you didn't think this was possible.

I'm admittedly a _touch_ annoyed with this, as I did a really in-depth amount of research and experimentation for this chapter including asking Vue experts that I trust in the VueLand Discord and seeking advise externally. Moreover, `useSlots` did not exist when I started writing this book 😅 

FWIW tho, my annoyance could just be because of a really rough month I've had thus far. Hiding this reply to not detract from the other/actual feedback.

**Comment by joezimjs** [comment]
> I'm currently investigating if `slots.default?.()?.length` has the same problems with children's `watch` and `watchEffect` that I have in my notes.

What problem is that? I wouldn't be terribly surprised if there are cases where this doesn't work because I'm using slots here in a way that I really haven't tried before and there really isn't any good documentation on these advanced uses of slots, which is a bit of a bummer.

> All that said, I authentically do appreciate the fact checking and potential corrections. I'll investigate and get back

No problem. I hope that it's all good so we can all be better informed and grow as developers.

**Comment by crutchcorn** [comment]
OK so I think I'm able to reproduce the behavior I originally ran into while researching that prevented me from mentioning this.

While your first example appears to run very well, you quickly run into issues when you expand on the idea:

```vue
<!-- App.vue -->
<script setup>
import Child from './Child.vue';
import ParentOfSlot from './ParentOfSlot.vue';
</script>

<template>
  <ParentOfSlot><Child /></ParentOfSlot>
</template>
```

```vue
<!-- Child.vue -->
<script setup>
import { watchEffect } from 'vue';

watchEffect(() => {
  alert('OH NO');
});
</script>

<template>
  <p>Child</p>
</template>
```

```vue
<!-- ParentOfSlot.vue -->
<script setup>
import { ref, useSlots, computed } from 'vue';
const slots = useSlots();

const defaultSlot = computed(() => slots.default?.());
</script>

<template>
  <slot />
  <p>There are {{ defaultSlot?.length ?? 0 }} items passed</p>
  <ul>
    <li v-for="child in defaultSlot"><component :is="child" /></li>
  </ul>
</template>
```

[StackBlitz](https://stackblitz.com/edit/vitejs-vite-aapbht?file=src%2FParentOfSlot.vue%2Csrc%2FApp.vue%2Csrc%2FChild.vue&terminal=dev)

Here, we're using your code in `ParentOfSlot` to project a `Child.vue` that has a `watchEffect` to trigger a simple `alert`:

```javascript
watchEffect(() => {
  alert('OH NO');
});
```

However, despite this `OH NO` only wanting to run once, it runs twice because the `defaultSlot` `VNode` is rendering twice:

- Once in `setup`
- Once in `template`

This duplicated effect occurs on **every** lifecycle method, not just `watchEffect`

Because of this behavioral issues with `useSlots`, I opted to not include it in my book.

**Comment by joezimjs** [comment]
@crutchcorn You're only getting that because you have `<slot />` at the beginning of your template. Remove that and it's fine.

**Comment by crutchcorn** [comment]
Hmm.... I was going to challenge this again, but my understanding of Angular's projected content is off and it turns out to behave the same as before. 

Re-opening, but will be adjusting the issue title and description

**Comment by joezimjs** [comment]
You're right that it probably doesn't belong in Fundamentals, but what you have right now makes it seem like Vue is impotent and considering React's explanation for "Passing Values to Projected Content" has a lengthy 5-part explanation, it kinda seems unfair to Vue that people are hearing that it "can't render children in a list". If you at least put something like "accomplishing this is a more advanced topic with caveats, therefore we will not be exploring it here" instead of "it can't", I'd be a little less sore about it. 

But, you do what you think is right; it's your book.
